src/core/search/search.cc

```diff
@@ -24,57 +25,154 @@ AstExpr ParseQuery(std::string_view query) {
   return driver.Take();
 }
 
+// Add results from matched to current, keep sorted and remove duplicates.
+void UnifyResults(vector<DocId>&& matched, vector<DocId>* current, vector<DocId>* tmp) {
+  swap(*current, *tmp);  // Move current to tmp so we can directly write into current
+  current->clear();
+  current->reserve(matched.size() + tmp->size());
+
+  merge(matched.begin(), matched.end(), tmp->begin(), tmp->end(), back_inserter(*current));
+  current->erase(unique(current->begin(), current->end()), current->end());
+}
+
+// Merge results from matched with current, keep sorted.
+void MergeResults(vector<DocId>&& matched, vector<DocId>* current, vector<DocId>* tmp) {
+  swap(*current, *tmp);  // Move current to tmp so we can directly write into current
+  current->clear();
+  current->reserve(matched.size() + tmp->size());
+
+  set_intersection(matched.begin(), matched.end(), tmp->begin(), tmp->end(),
+                   back_inserter(*current));
+}
+
 struct BasicSearch {
-  bool Check(monostate, string_view) {
-    return false;
+  // Get casted sub index by field
+  template <typename T> T* GetIndex(string_view field) {
+    static_assert(is_base_of_v<BaseIndex, T>);
+    auto index = indices->GetIndex(field);
+    DCHECK(index);  // TODO: handle not existing erorr
+    auto* casted_ptr = dynamic_cast<T*>(index);
```

Comment:
dynamic_cast is cheating :)

Reply:
It can be of multiple types 💁🏻There are two options, variants and virtual pointers. 

Virtual casting is a perfectly fine and safe cast.

Reply:
it uses RTTI, no?
also, it throws exceptions that we do not support.

Reply:
anyway, it's fine, it's just my personal taste.

Reply:
Yes, it uses RTTI. Like every virtual method call. 

It throws only if the pointer is invalid. If the cast failed, it returns nullptr.

Reply:
it's not accurate. rtti is special. See https://gcc.gnu.org/onlinedocs/gcc-4.4.7/gcc/C_002b_002b-Dialect-Options.html

```
Disable generation of information about every class with virtual functions for use by the C++ runtime type identification features (`dynamic_cast' and `typeid'). If you don't use those parts of the language, you can save some space by using this flag.
```

Reply:
Passer-by comment:

RTTI is somewhat different (think of it as an extension) to virtual calls.
It's true that virtual calls use a vtable, with a (small) cost of double dereference for invoking methods. However RTTI requires additional information (which can be disabled when not affecting virtual calls).
Disabling RTTI can save in binary size (I can't estimate but I would imagine it's negligible at least in our case). Also RTTI casts (i.e. casting down using `dynamic_cast`) is slower than regular casts or even virtual calls.

Again, I don't think it should not be used :)

There's some more info here for the curious: https://shaharmike.com/cpp/vtable-part1/